code/components/gta-core-five/src/PoolManagement.cpp

```diff
@@ -237,15 +401,131 @@ static void* PoolAllocateWrap(atPoolBase* pool)
+
+				int32_t currentSize = pool->m_count;
+				int32_t newSize = currentSize + data.m_growthStep;
+
+				// make sure we don't overflow
+				if (newSize > targetMax)
+				{
+					newSize = targetMax;
+				}
+					
+				trace("resizing pool %p [%p]: (%s) 0x%x -> 0x%x\n", (void*)pool, (void*)pool->m_data, poolEntries.LookupHash(g_inversePools[pool]), currentSize, newSize);
+
+				if (newSize == targetMax)
+				{
+					trace("\t>>> WARNING: pool is at max capacity!\n", (void*)pool);
+				}
+
+				int32_t newDelta = newSize - currentSize;
+				assert(newDelta > 0);
+
+				// add new memory 
+				void* res = VirtualAlloc(pool->m_data + (currentSize * poolEntrySize), newDelta * poolEntrySize, MEM_COMMIT, PAGE_READWRITE);
+				assert(res);
+
+				res = VirtualAlloc(pool->m_flags + currentSize, newDelta, MEM_COMMIT, PAGE_READWRITE);
+				assert(res);
+
+				// every new entry is free and is the first generation
+				for (int32_t i = currentSize; i < newSize; ++i)
+				{
+					pool->m_flags[i].isFree = 1;
+					pool->m_flags[i].generation = 1;
+				}
+
+				// populate free list
+				int32_t* freeList = &pool->m_freeList;
+				assert(*freeList == -1);
+
+				for (int32_t i = currentSize; i < newSize; ++i)
+				{
+					*freeList = i;
+					freeList = (int32_t*)&pool->m_data[i * poolEntrySize];
+				}
+				*freeList = -1;
+
+				assert(freeList != &pool->m_freeList);
+
+				pool->m_count = newSize;
+				pool->m_lastAlloc = newSize - 1;
+
+				// try again
+				value = g_origPoolAllocate(pool);
+				assert(value);
+
+				// asset stores have a dead-simple open address hashmap tacked on,
+				// which we need to resize accordingly as well
+				if (auto it = g_hackedHashmaps.find(pool); it != g_hackedHashmaps.end())
+				{
+					auto* assetMap = it->second;
+					trace("[in resize] hashmap with alloc count %zu, modulus %zu, and free idx %d!\n", assetMap->allocCount, assetMap->modulus, assetMap->nextFreeEntry);
+					
+					int32_t nextFree = assetMap->nextFreeEntry;
+
+					auto* alloc = rage::GetAllocator();
+
+					AssetStoreHashMap::Bucket* newBuckets = (AssetStoreHashMap::Bucket*)alloc->Allocate(sizeof(AssetStoreHashMap::Bucket) * newSize, alignof(AssetStoreHashMap::Bucket), 0);
+					memcpy(newBuckets, assetMap->data, currentSize * sizeof(AssetStoreHashMap::Bucket));
+					rage::GetAllocator()->Free(assetMap->data);
+					assetMap->data = newBuckets;
+
+					/* TODO: resize if we want to make modulus larger
+					uint32_t* newPrimary = (uint32_t*)alloc->Allocate(sizeof(uint32_t) * newSize, alignof(uint32_t), 0);
+					memcpy(newPrimary, assetMap->primary, currentSize * sizeof(uint32_t));
+					rage::GetAllocator()->Free(assetMap->primary);
+					assetMap->primary = newPrimary;
+					*/
+
+					// populate new free entries in hashmap
+					if (nextFree == -1)
+					{
+						assetMap->nextFreeEntry = currentSize;
+						nextFree = currentSize;
+					}
+					else
+					{
+						assetMap->data[nextFree].link = currentSize;
```

Comment:
todo: pretty sure this is wrong,`nextFree = currentSize;` needs to be on both branches